### PR TITLE
Merging Enumerable Methods code

### PR DESCRIPTION
**Pull request by helman101**
## What?
Enumerable module provides lots of handy iteration methods in ruby. We rebuild the following methods : 

- each
- each_with_index
- select
- all
- any
- none
- count
- map
- inject

## Why?
Part of Ruby Project to write Enumerable methods code.

## How?
Mentioned in Readme

## Testing?
- Run the code using the demo link in readme.
- Change the inputs to check if necessary to check for working

### CODE REVIEW

File: enum_methods.rb [hunk 1]
```diff
@@ -0,0 +1,146 @@
+module Enumerable
+  def my_each
+    if block_given?
+      for i in self
+        yield(i)
+      end
+    else
+      puts 'No Block Given'
+    end
+  end
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x] my_each returns an Enumerator if no block is given.

File: enum_methods.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    count = 0
+    if block_given?
+      for i in self
+        yield(i, count)
+        count += 1
+      end
+    else
+      puts 'No Block Given'
+    end
+  end
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x] my_each_with_index returns an enumerator if no block is given

File: enum_methods.rb [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    if instance_of?(Hash)
+      result = {}
+      my_each { |key, value| result[key] = value if yield(key, value) == true }
+    else
+      result = []
+      my_each { |value| result.push(value) if yield(value) == true }
+    end
+    result
+  end
+
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x] my_select returns an enumerator if no block is given

File: enum_methods.rb [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  def my_all?
+    count = 0
+    my_each { |value| count += 1 if yield(value) == true }
+    count == length
+  end
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x]  my_all? returns false if the block returns false or nil::range

- [x]  my_all? when no block or argument is given returns true when none of the collection members are false or nil

- [x] my_all? when no block or argument is given returns false when one of the collection members are false or nil

- [x] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer

- [x] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric

File: enum_methods.rb [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+  def my_any?
+    count = 0
+    my_each { |value| break count += 1 if yield(value) == true }
+    count.positive?
+  end
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil

- [x] my_any? when no block or argument is given returns false if at least one of the collection is not true

- [x] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric

- [x] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer

- [x] my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex

- [x] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex

- [x] my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

File: enum_methods.rb [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  def my_none?
+    count = 0
+    my_each { |value| break count += 1 if yield(value) == true }
+    !count.positive?
+  end
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x] my_none? when no block or argument is given returns false only if one of the collection members is true

- [x] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class

- [x] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric

- [x] my_none? when a class is passed as an argument returns false if any of the collection is a member of such class

- [x] my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
      Failure/Error: expect(words.my_none?(/z/)).to be words.none?

File: enum_methods.rb [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    if block_given?
+      my_each do |value|
+        count += 1 if yield(value) == true
+      end
+      count
+    else
+      puts 'No Block Given'
+    end
+  end
+
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x] my_count returns the number of items in enum through enumeration

- [x] my_count returns the number of items in enum through enumeration::range

- [x] my_count counts the number of items in enum that are equal to item if an argument is given

File: enum_methods.rb [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        my_each { |key, value| result.push(yield(key, value)) }
+        result
+      else
+        self
+      end
+    else
+      my_each { |key, value| result.push(proc.call(key, value)) }
+      result
+    end
+  end
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x] my_map returns an Enumerator if no block is given

File: enum_methods.rb [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+          arg = yield(arg, self[count])
+          count += 1
+        end
+      end
+    else
+      my_each { |value| arg = yield(arg, value) }
+    end
+    arg
+  end
+end
```
Comment (mahmoud717) [inline enum_methods.rb]: - [x] my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range

- [x] my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block

- [x] my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method

- [x] my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range

- [x] my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method

- [x] my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @helman101 & @akshay-narkar**

Great job with the project so far, but there are still some changes to be made:
 
---
### Requested 🔴 


- [x] please move all of your test cases to a test file

- [x] check the inline comments.


---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Review by PhillipUg — CHANGES_REQUESTED** [review]
Hey team, 👋

You have done a great job on this project so far, however, there still some changes to be made before we can approve this request 👇🏿 

### Changes Required ♻️ 

- [x] Your `my_all`, `my_map` and `my_none` methods are failing. I am going to provide you with some test cases below. make sure these methods pass all the provided test cases.

```
# 4. my_all? (example test cases)
puts 'my_all?'
puts '-------'
p [3, 5, 7, 11].my_all?(&:odd?) # => true
p [-8, -9, -6].my_all? { |n| n < 0 } # => true
p [3, 5, 8, 11].my_all?(&:odd?) # => false
p [-8, -9, -6, 0].my_all? { |n| n < 0 } # => false
# test cases required by tse reviewer
p [1, 2, 3, 4, 5].my_all? # => true
p [1, 2, 3].my_all?(Integer) # => true
p %w[dog door rod blade].my_all?(/d/) # => true
p [1, 1, 1].my_all?(1) # => true
puts

# 6. my_none? (example test cases)
puts 'my_none?'
puts '--------'
p [3, 5, 7, 11].my_none?(&:even?) # => true
p %w[sushi pizza burrito].my_none? { |word| word[0] == 'a' } # => true
p [3, 5, 4, 7, 11].my_none?(&:even?) # => false
p %w[asparagus sushi pizza apple burrito].my_none? { |word| word[0] == 'a' } # => false
# test cases required by tse reviewer
p [1, 2, 3].my_none? # => false
p [1, 2, 3].my_none?(String) # => true
p [1, 2, 3, 4, 5].my_none?(2) # => false
p [1, 2, 3].my_none?(4) # => true
puts

# 8. my_map
puts 'my_map'
puts '------'
p [1, 2, 3].my_map { |n| 2 * n } # => [2,4,6]
p %w[Hey Jude].my_map { |word| word + '?' } # => ["Hey?", "Jude?"]
p [false, true].my_map(&:!) # => [true, false]
my_proc = proc { |num| num > 10 }
p [18, 22, 5, 6].my_map(my_proc) { |num| num < 10 } # => true true false false
puts
```

- [x] You have taken the `multiply_els` method to the `tests.rb` file. it should be in the `enum_methods.rb` file but outside the module. It is part of the methods you're required to implement. 

#### Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack as **Phillip Musiime**
I would be happy to help! 👩‍💻 

Happy coding 🤓

**Comment by mimipeshy** [comment]
## STATUS: CHANGES REQUIRED :recycle: 

Hey team :wave:  :wave: , great job in implementing all changes given by TSE's. However, let's fix one more failing test

- [x] enumerables #my_inject should raise a "LocalJumpError" when no block or argument is given

That's all :relieved:  , please submit another review when done. 

Regards
Peris Ndanu
Feel free to slack me anytime if you have any questions (perisndanu).

**Review by umairarshadbutt — APPROVED** [review]
# **Status: PROJECT APPROVED** :champagne:
Excellent job on this project guys. :fireworks: Keep it up.
##
**All the required tasks have been fulfilled in this project. I will go ahead and mark this as complete.**
- Congratulations on the splendid job @helman101 and @akshay-narkar ! :+1: 
- You can now merge to master and submit the project completion form :rocket: . I wish you all the very best in your next tasks :1st_place_medal:
##
**If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Pull request merged**